ambry-store/src/test/java/com.github.ambry.store/BlobStoreStatsTest.java

```diff
@@ -241,8 +315,351 @@ public void testValidDataSizeAfterDeletes() throws InterruptedException, StoreEx
+    throttlers.put(BlobStoreStats.IO_SCHEDULER_JOB_TYPE, mockThrottler);
+    BlobStoreStats blobStoreStats =
+        setupBlobStoreStats(2 * (int) (state.time.milliseconds() / BUCKET_SPAN_IN_MS), state.time.milliseconds());
+    // proceed only when the scan is started
+    assertTrue("IndexScanner took too long to start", scanStartedLatch.await(5, TimeUnit.SECONDS));
+    advanceTimeToNextSecond();
+    verifyAndGetLogSegmentValidSize(blobStoreStats, new TimeRange(state.time.milliseconds(), 0L));
+    verifyAndGetContainerValidSize(blobStoreStats, state.time.milliseconds());
+    int throttleCountBeforeRequests = mockThrottler.throttleCount.get();
+    // add 3 put with expiry
+    long expiresAtInMs = state.time.milliseconds() + TimeUnit.SECONDS.toMillis(30);
+    List<IndexEntry> newPutEntries = state.addPutEntries(3, CuratedLogIndexState.PUT_RECORD_SIZE, expiresAtInMs);
+    for (IndexEntry entry : newPutEntries) {
+      blobStoreStats.handleNewPutEntry(entry.getValue());
+    }
+    // delete the first new put with expiry
+    MockId firstPutWithExpiry = getIdToDelete(newPutEntries.get(0).getKey());
+    newDelete(blobStoreStats, firstPutWithExpiry);
+    // a probe put with a latch to inform us about the state of the queue
+    blobStoreStats.handleNewPutEntry(new MockIndexValue(queueProcessedLatch, state.index.getCurrentEndOffset()));
+    assertTrue("QueueProcessor took too long to process the new entries",
+        queueProcessedLatch.await(5, TimeUnit.SECONDS));
+    advanceTimeToNextSecond();
+    verifyAndGetLogSegmentValidSize(blobStoreStats, new TimeRange(state.time.milliseconds(), 0L));
+    verifyAndGetContainerValidSize(blobStoreStats, state.time.milliseconds());
+    // add 3 put with no expiry
+    newPutEntries = state.addPutEntries(3, CuratedLogIndexState.PUT_RECORD_SIZE, Utils.Infinite_Time);
+    for (IndexEntry entry : newPutEntries) {
+      blobStoreStats.handleNewPutEntry(entry.getValue());
+    }
+    // delete one of the newly added put
+    newDelete(blobStoreStats, state.getIdToDeleteFromIndexSegment(state.referenceIndex.lastKey()));
+    queueProcessedLatch = new CountDownLatch(1);
+    // a probe put with a latch to inform us about the state of the queue
+    blobStoreStats.handleNewPutEntry(new MockIndexValue(queueProcessedLatch, state.index.getCurrentEndOffset()));
+    assertTrue("QueueProcessor took too long to process the new entries",
+        queueProcessedLatch.await(5, TimeUnit.SECONDS));
+    advanceTimeToNextSecond();
+    verifyAndGetLogSegmentValidSize(blobStoreStats, new TimeRange(state.time.milliseconds(), 0L));
+    verifyAndGetContainerValidSize(blobStoreStats, state.time.milliseconds());
+    // advance time beyond expiration of the blobs and verify no double counting for expiration and delete
+    long timeToLiveInMs = expiresAtInMs - state.time.milliseconds() < 0 ? 0 : expiresAtInMs - state.time.milliseconds();
```

Comment:
many entries are added (time is advanced when index segments roll over) between the time `expiresAtInMs` is  taken and the time `timeToLiveInMs` is calculated.